ts/Stock/Indicators/Klinger/KlingerIndicator.ts

```diff
@@ -0,0 +1,434 @@
+         */
+        function isLengthValid(series: LineSeries): (boolean|undefined) {
+            return series && series.xData && series.xData.length >=
+                (options.params as any).slowAvgPeriod;
+        }
+
+        return !!(isLengthValid(series) && isLengthValid(volumeSeries) && isSeriesOHLC);
+    }
+
+    public getCM(
+        previousCM: number,
+        DM: number,
+        trend: number,
+        previousTrend: number,
+        prevoiusDM: number
+    ): number {
+        let CM;
+        if (trend === previousTrend) {
+            CM = previousCM + DM;
+        } else {
+            CM = prevoiusDM + DM;
+        }
+        return correctFloat(CM);
+    }
+
+    public getDM(
+        high: number,
+        low: number
+    ): number {
+        return correctFloat(high - low);
+    }
+
+    public getVolumeForce(yVal: Array<Array<number>>): Array<Array<number>> {
+        const volumeForce: Array<Array<number>> = [];
+
+        let CM: number = 0, // cumulative measurement
+            DM: number, // daily measurement
+            force: number,
+            i: number = 1, // start from second point
+            previousCM: number = 0,
+            previousDM: number = yVal[0][1] - yVal[0][2], // initial DM
+            previousTrend: number = 0,
+            trend: number;
+
+        for (i; i < yVal.length; i++) {
+            trend = this.calculateTrend(yVal, i);
+            DM = this.getDM(yVal[i][1], yVal[i][2]);
+            // For the first iteration when the previousTrend doesn't exist,
+            // previousCM doesn't exist either, but it doesn't matter becouse
+            // it's filltered out in the getCM method in else statement,
+            // (in this iteration, previousCM can be raplaced with the DM).
+            CM = this.getCM(previousCM, DM, trend, previousTrend, previousDM);
+
+            force = (this.volumeSeries.yData as any)[i] * trend * Math.abs(2 * ((DM / CM) - 1)) * 100;
+            volumeForce.push([force]);
+
+            // Before next iteration, assign the current as the previous.
+            previousTrend = trend;
+            previousCM = CM;
+            previousDM = DM;
+        }
+        return volumeForce;
+    }
+
+    public getEMA(
+        yVal: (Array<number>|Array<Array<number>>),
+        prevEMA: (number|undefined),
+        SMA: number,
+        EMApercent: number,
+        index?: number,
+        i?: number,
+        xVal?: Array<number>
+    ): [number, number] {
+
+        return EMAIndicator.prototype.calculateEma(
+            xVal || [],
+            yVal,
+            typeof i === 'undefined' ? 1 : i,
+            EMApercent,
+            prevEMA,
+            typeof index === 'undefined' ? -1 : index,
+            SMA
+        );
+    }
+
+    public getSMA(
+        period: number,
+        index: number,
+        values: Array<Array<number>>
+    ): number {
+
+        return EMAIndicator.prototype.accumulatePeriodPoints(period, index, values) / period;
```

Comment:
I'm a bit confused here. The method is called `getSMA()` but returns the sum of points (`accumulatePeriodPoints` returns the sum, not the average).

Reply:
It is divided by `period`. However, it exceeds 80 characters, so maybe we've finally found a case where too long lines can be confusing :D

Reply:
Oh boy.. Indeed, there's `/ period` at the end 😅